README
Crypt-PerfectPaperPasswords version 0.05

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

Alternatively, to install with Module::Build, you can use the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

DEPENDENCIES

None.

COPYRIGHT AND LICENCE

Copyright (C) 2007, Andy Armstrong

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Crypt-PerfectPaperPasswords

0.01    2007-11-17
        - Initial release.

0.02    2007-11-17
        - Oops - included missing MANIFEST items

0.03    2007-11-17
        - Removed BigInt silliness

0.04    2007-11-17
        - Sigh. Fixed t/000-load.t

0.05    2007-11-17
        - pppp command line tool working
        - Optional Math::BigInt support for big keys

0.06    2007-11-17


Makefile.PL
use strict;
#use warnings;
use ExtUtils::MakeMaker;

eval 'use ExtUtils::MakeMaker::Coverage';

WriteMakefile(
    ( MM->can( 'signature_target' ) ? ( SIGN => 1 ) : () ),
    NAME          => 'Crypt::PerfectPaperPasswords',
    AUTHOR        => 'Andy Armstrong <andy@hexten.net>',
    LICENSE       => 'perl',
    VERSION_FROM  => 'lib/Crypt/PerfectPaperPasswords.pm',
    ABSTRACT_FROM => 'lib/Crypt/PerfectPaperPasswords.pm',
    PL_FILES      => {},
    PREREQ_PM     => {
        'Test::More'      => 0,
        'Crypt::Rijndael' => 0,
        'Digest::SHA256'  => 0,
        'Getopt::Long'    => 0,
        'Scalar::Util'    => 0,
        'Time::HiRes'     => 0,
        'Getopt::Long'    => 0,
        'Pod::Usage'      => 0,
    },
    EXE_FILES => ['bin/pppp'],
    dist      => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Crypt-PerfectPaperPasswords-*' },
);


lib/App/PerlPPP.pm
package App::PerlPPP;

use strict;
use warnings;
use Carp;
use Getopt::Long;
use Crypt::PerfectPaperPasswords;
use Pod::Usage;

=head1 NAME

App::PerlPPP - Command line tool for Perfect Paper Passwords

=head1 VERSION

This document describes App::PerlPPP version 0.06

=cut

our $VERSION = '0.06';

=head1 SYNOPSIS

    use App::PerlPPP;

    my $app = App::PerlPPP->new;
    $app->parse_args(@ARGV);
    $app->run;

=head1 DESCRIPTION

=head1 INTERFACE 

=head2 C<< new >>

=cut

{
    my %ARG_SPEC;

    BEGIN {

        sub _range_spec {
            my ( $name, $default, $min, $max ) = @_;
            return [
                $default,
                sub {
                    my ( $self, $value ) = @_;
                    die "$name must be between $min and $max\n"
                      if $value < $min || $value > $max;
                    return $value;
                  }
            ];
        }

        sub _array_spec {
            return [
                [],
                sub {
                    my $self = shift;
                    return [ map { 'ARRAY' eq ref $_ ? @$_ : $_ } @_ ];
                },
            ];
        }

        %ARG_SPEC = (
            show_man   => [0],
            show_help  => [0],
            args       => _array_spec(),
            columns    => _range_spec( 'columns', 7, 1, 100 ),
            rows       => _range_spec( 'rows', 10, 1, 100 ),
            codelen    => [undef],
            passphrase => [undef],
            alphabet   => [undef],
            key        => [
                undef,
                sub {
                    my ( $self, $key ) = @_;
                    die "Key must be 64 characters long\n"
                      unless length( $key ) == 64;
                    die "Key must be hexadecimal (0-9, A-F)\n"
                      unless $key =~ /^[0-9A-F]{64}$/i;
                    return $key;
                },
            ],
            title => ['Perfect Paper Passwords'],
        );

        while ( my ( $name, $spec ) = each %ARG_SPEC ) {
            no strict 'refs';
            my $validator = $spec->[1] || sub { shift; shift };
            *{ __PACKAGE__ . '::' . $name } = sub {
                my $self = shift;
                $self->{$name} = $self->$validator( @_ )
                  if ( @_ );
                my $value = $self->{$name};
                return ( wantarray && 'ARRAY' eq ref $value )
                  ? @$value
                  : $value;
            };
        }
    }

    sub new {
        my ( $class, %args ) = @_;

        my $self = bless {}, $class;

        while ( my ( $name, $spec ) = each %ARG_SPEC ) {
            my $value
              = exists $args{$name} ? delete $args{$name} : $spec->[0];
            $self->$name( $value )
              if defined $value;
        }

        croak "Unknown options: ", join( ', ', sort keys %args )
          if keys %args;

        return $self;
    }
}

=head2 C<< args >>

=head2 C<< alphabet >>

=head2 C<< codelen >>

=head2 C<< columns >>

=head2 C<< rows >>

=head2 C<< key >>

=head2 C<< passphrase >>

=head2 C<< title >>

=head2 C<< show_help >>

=head2 C<< show_man >>

=head2 C<< parse_args >>

=cut

sub parse_args {
    my ( $self, @args ) = @_;

    local @ARGV = @args;

    my %options;

    GetOptions(
        'help|?'       => \$options{show_help},
        man            => \$options{show_man},
        'key=s'        => \$options{key},
        'passphrase=s' => \$options{passphrase},
        'columns=i'    => \$options{columns},
        'rows=i'       => \$options{rows},
        'title=s'      => \$options{title},
        'alphabet=s'   => \$options{alphabet},
        'codelen=i'    => \$options{codelen},
    ) or pod2usage();

    while ( my ( $name, $value ) = each %options ) {
        $self->$name( $value ) if defined $value;
    }

    $self->args( @ARGV );
}

=head2 C<< run >>

=cut

sub run {
    my $self = shift;

    if ( $self->show_help ) {
        $self->do_help;
    }
    elsif ( $self->show_man ) {
        pod2usage( -verbose => 2, -exitstatus => 0 );
    }
    else {
        my @args = $self->args;
        pod2usage() unless @args;
        my $verb = shift @args;
        if ( my $code = $self->can( "do_$verb" ) ) {
            $self->$code( @args );
        }
        else {
            die "Unknown action: $verb\n";
        }
    }
}

=head1 ACTIONS

=head2 C<< do_card >>

Output a card

=cut

sub do_card {
    my ( $self, @args ) = @_;
    my $card_no = @args ? shift @args : 1;
    die "Card numbers start at 1\n" if $card_no < 1;

    my $title = $self->title;
    my $rows  = $self->rows;
    my $cols  = $self->columns;
    my $ppp   = $self->_make_ppp;
    my @passcodes
      = $ppp->passcodes( $card_no, $rows * $cols, $self->_get_key );

    my $colw   = length( $passcodes[0] );
    my $center = sub {
        my $str = shift;
        my $pad = $colw - length $str;
        return $str if $pad <= 0;
        return ( ' ' x ( $pad / 2 ) ) . $str
          . ( ' ' x ( ( $pad + 1 ) / 2 ) );
    };

    my $row_fmt = "%4d";
    my @hdr     = ( ' ' x ( length sprintf( $row_fmt, 1 ) ) );
    my $col     = 'A';
    push @hdr, $center->( $col++ ) for ( 1 .. $cols );
    my $hdr = join( ' ', @hdr );
    my $rule = '=' x length $hdr;
    print "$title [$card_no]\n$rule\n$hdr\n$rule\n";
    for ( 1 .. $rows ) {
        print join( ' ',
            sprintf( $row_fmt, $_ ),
            splice( @passcodes, 0, $cols ) ),
          "\n";
    }
    print "$rule\n";
}

=head2 C<do_newkey>

Create and display a new random key.

=cut

sub do_newkey {
    my $self = shift;
    if ( my $key = $self->_make_key ) {
        print "Specified key is $key\n";
    }
    else {
        print "Generated key is ",
          $self->key( $self->_make_ppp->random_sequence ), "\n";
    }
}

=head2 C<do_help>

Output help page

=cut

sub do_help {
    my $self = shift;
    pod2usage( -verbose => 1 );
}

sub _make_key {
    my $self = shift;

    if ( defined( my $key = $self->key ) ) {
        return $key;
    }
    elsif ( defined( my $phrase = $self->passphrase ) ) {
        return $self->key(
            $self->_make_ppp->sequence_from_key( $phrase ) );
    }
    else {
        return;
    }
}

sub _get_key {
    my $self = shift;
    return $self->_make_key
      || die "Must supply --key or --passphrase\n";
}

sub _make_ppp {
    my $self = shift;
    my %args;
    for my $a ( qw( alphabet codelen ) ) {
        if ( defined( my $value = $self->$a() ) ) {
            $args{$a} = $value;
        }
    }

    return $self->{_ppp} ||= Crypt::PerfectPaperPasswords->new( %args );
}

1;
__END__

=head1 CONFIGURATION AND ENVIRONMENT
  
App::PerlPPP requires no configuration files or environment variables.

=head1 DEPENDENCIES

L<Crypt::PerfectPaperPasswords>

=head1 INCOMPATIBILITIES

None reported.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-crypt-perfectpaperpasswords@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.

=head1 AUTHOR

Andy Armstrong  C<< <andy@hexten.net> >>

Original Perfect Paper Passwords implementation by Steve Gibson. More details
here:

    http://www.grc.com/ppp.htm

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2007, Andy Armstrong C<< <andy@hexten.net> >>.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


misc/PPPinPHP_11-02-07/ppp.php
<?php

// PPP in PHP (November 2, 2007)
// An implementation of the PPP CryptoSystem designed by Steve Gibson of GRC
// Full details and spec can be found at http://www.grc.com/ppp

// Implementation in PHP by Bob Somers (www.bobsomers.com)
// Tested on PHP version 5.2.3

// REQUIREMENTS:
// + PHP5, preferably 5.2.3 or higher (that's what I tested with)
// + The hash() function must support the SHA256 hash. You can check this with the hash_algos() function.
// + The bcmath extension for dealing with numbers of arbitrary size.
// + The mcrypt extension (built against libmcrypt > 2.4.x) for 128-bit Rijndael.

// define the passcode alphabet
$PC_ALPHABET = "23456789!@#%+=:?abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPRSTUVWXYZ";

// generates a 256-bit sequence key by hashing the passed string with SHA256
// it isn't recommended to use this method, rather, you should generate a random
// key with GenerateRandomSequenceKey() instead
// returns the sequence key as a hex string
function GenerateSequenceKeyFromString($passphrase)
{
	return hash('sha256', $passphrase);
}

// generates a random 256-bit sequence key
// returns the sequence key as a hex string
function GenerateRandomSequenceKey()
{
	$randomness = get_loaded_extensions();
	$randomness[] = php_uname();
	$randomness[] = memory_get_usage();
	$randomness = implode(microtime(), $randomness);
	return hash('sha256', $randomness);
}

// returns an array of $count passcodes starting with the passcode number $first,
// generated using the passed sequence key
// the $first and $count variables should be passed as arbitrary-length decimal strings (see the bcmath functions)
// the $sequence_key should be passed as a hex string
function RetrievePasscodes($first, $count, $sequence_key)
{
	global $PC_ALPHABET; // bring in the passcode alphabet

	$passcodes = array();
	
	// pack our hex string sequence key into a binary string
	$sk = pack("H*", $sequence_key);
	
	// because we need to handle these as aribitrary position numbers, we need to write a for loop using
	// bcmath functions (as a while loop), which is why this looks bizarre
	$i = $first;
	$limit = bcadd($first, $count);
	while (bccomp($i, $limit) < 0)
	{
		$pc_start = bcmul($i, '24'); // how many bits into the sequence does the passcode start?
		$n = bcdiv($pc_start, '128', 0); // what 128-bit block does that start in? (in other words, what's the value of our counter?)
		$offset = bcmod($pc_start, '128'); // how many bits into the n'th 128-bit block does the passcode start?
		
		if (bccomp(bcsub('128', $offset), '24') < 0)
		{
			// we don't have enough bits left in a single cipher, we're going to need to calculate two
			$ciphers = array();
			for ($j = 0; $j < 2; $j++)
			{
				$n_bits = pack("V*", bcadd($n, $j));
				error_reporting(error_reporting() & ~E_WARNING); // mcrypt complains that we aren't using an initialization vector
				$enc_bits = mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $sk, $n_bits, MCRYPT_MODE_ECB);
				error_reporting(error_reporting() | E_WARNING); // we still want to get warnings, though
				$enc_hex = unpack("H*", $enc_bits);
				$ciphers[$j] = $enc_hex[1];
			}
			
			// grab the chunks we need from each cipher and combine them
			// note that $offset will always be less than 128, so we don't need to use bcmath functions here
			$first_chunk = substr($ciphers[0], $offset / 4, (128 - $offset) / 4);
			$second_chunk = substr($ciphers[1], 0, (24 - (128 - $offset)) / 4);
			$pc_hex = $first_chunk . $second_chunk;
		}
		else
		{
			// a single cipher will do the trick
			$n_bits = pack("V*", $n);
			error_reporting(error_reporting() & ~E_WARNING); // mcrypt complains that we aren't using an initialization vector
			$enc_bits = mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $sk, $n_bits, MCRYPT_MODE_ECB);
			error_reporting(error_reporting() | E_WARNING); // we still want to get warnings, though
			$enc_hex = unpack("H*", $enc_bits);
			$enc_hex = $enc_hex[1];
			
			// grab the chunk which represents our passcode
			$pc_hex = substr($enc_hex, $offset / 4, 6);
		}
		
		// reverse byte order, regroup into 6-bit chunks, and translate to the passcode alphabet
		$pc_bin = str_pad(decbin(hexdec($pc_hex)), 24, "0", STR_PAD_LEFT);
		$pc_chars = array();
		for ($j = 0; $j < 3; $j++)
		{
			$pc_chars[$j] = substr($pc_bin, -8 * ($j + 1), 8);
		}
		$pc_bin = implode("", $pc_chars);
		$pc_chars = array();
		for ($j = 0; $j < 4; $j++)
		{
			$pc_chars[$j] = $PC_ALPHABET[bindec(substr($pc_bin, $j * 6, 6))];
		}
		$passcode = implode("", array_reverse($pc_chars));
		
		// add this passcode to our list of passcodes to return
		$passcodes[] = $passcode;
		
		// increment our "for" loop
		$i = bcadd($i, '1');
	}
	
	return $passcodes;
}

?>

lib/Crypt/PerfectPaperPasswords.pm
package Crypt::PerfectPaperPasswords;

use warnings;
use strict;
use Carp;
use Crypt::Rijndael;
use Digest::SHA256;
use Time::HiRes qw(time);
use Scalar::Util qw(refaddr);

=head1 NAME

Crypt::PerfectPaperPasswords - Steve Gibson's Perfect Paper Passwords

=head1 VERSION

This document describes Crypt::PerfectPaperPasswords version 0.06

=cut

our $VERSION = '0.06';

=head1 SYNOPSIS

    use Crypt::PerfectPaperPasswords;

    my $pass_phrase  = 'Fromage';
    my $ppp          = Crypt::PerfectPaperPasswords->new;
    my $sequence_key = $ppp->sequence_from_key( $pass_phrase );
    my $first        = 1;
    my $count        = 100;
    my @passcodes    = $ppp->passcodes( $first, $count, $sequence_key );

=head1 DESCRIPTION

From L<https://www.grc.com/ppp.htm>

    GRC's "Perfect Paper Passwords" (PPP) system is a straightforward,
    simple and secure implementation of a paper-based One Time Password
    (OTP) system. When used in conjunction with an account name &
    password, the individual "passcodes" contained on PPP's "passcards"
    serve as the second factor ("something you have") of a secure multi-
    factor authentication system.

This is a Perl implementation of the PPP passcode generator.

=head1 INTERFACE 

=head2 C<< new >>

Create a new C<Create::PerfectPaperPasswords> instance. Options may
be passed:

    my $ppp = Crypt::PerfectPaperPasswords->new(
        alphabet => '0123456789abcdef',
        codelen  => 2
    );

The following options are supported:

=over

=item C<alphabet>

The alphabet to use for encoding. Defaults to Steve Gibson's:

    23456789!@#%+=:?abcdefghijkmnopq
    rstuvwxyzABCDEFGHJKLMNPRSTUVWXYZ

The size of the alphabet need not be a power of two.

=item C<codelen>

The number of raw bytes in each passcode. You must have L<Math::BigInt>
installed to handle values greater than 4.

=back

=cut

{
    my %DEFAULT_ARGS;

    BEGIN {
        %DEFAULT_ARGS = (
            alphabet => '23456789!@#%+=:?'
              . 'abcdefghijkmnopqrstuvwxyz'
              . 'ABCDEFGHJKLMNPRSTUVWXYZ',
            codelen => 3,
        );

        for my $method ( keys %DEFAULT_ARGS ) {
            no strict 'refs';
            *{ __PACKAGE__ . '::' . $method } = sub {
                my $self = shift;
                croak "Can't set $method" if @_;
                return $self->{$method};
            };
        }
    }

    sub new {
        my $class = shift;
        my %args = ( %DEFAULT_ARGS, @_ );

        my $alphabet = delete $args{alphabet};

        croak "Alphabet must be at least two characters long"
          unless length( $alphabet ) >= 2;

        my %got = ();
        $got{$_}++ for split //, $alphabet;
        my @dups = sort grep { $got{$_} > 1 } keys %got;
        croak "Duplicate characters in alphabet: ", join( ', ', @dups )
          if @dups;

        my $codelen = delete $args{codelen};

        croak "Code length must be between 1 and 32"
          if $codelen < 1 || $codelen > 32;

        if ( $codelen > 4 && !_got_bigint() ) {
            croak
              "Please install Math::BigInt to handle code lengths > 4";
        }

        my $self = bless {
            alphabet => $alphabet,
            codelen  => $codelen,
            seed     => time(),
        }, $class;

        croak "Unknown options: ", join( ', ', sort keys %args ), "\n"
          if keys %args;

        return $self;
    }
}

=head2 C<< alphabet >>

Get the alphabet used by this object.

    my $alphabet = $ppp->alphabet;

=head2 C<< codelen >>

Get the code length for this object.

    my $codelen = $ppp->codelen;

=head2 C<< sequence_from_key >>

Generate a sequence key from a passphrase.

    my $seq_key = $ppp->sequence_from_key( 'Fromage' );

=cut

sub sequence_from_key {
    my $self = shift;
    my $key  = shift;

    my $sha = Digest::SHA256::new( 256 );
    $sha->add( $key );
    my $digest = $sha->hexdigest;
    $digest =~ s/\s+//g;
    return $digest;
}

=head2 C<< random_sequence >>

Generate a random sequence key.

    my $seq_key = $ppp->random_sequence;

Relies on the output of C<random_data> for its entropy.

=cut

sub random_sequence {
    my $self = shift;
    return $self->sequence_from_key( $self->random_data );
}

=head2 C<< random_data >>

Returns some random data. This is the entropy source for
C<random_sequence>. This implementation returns a string
that is the concatenation of

=over

=item * The real time (using the microsecond clock)

=item * The next seed value

=item * Address of C<$self>

=item * Address of a newly allocated scalar

=item * Process ID

=back

The seed value is the microsecond time when this object was created and
is incremented by one each time it's used.

For a lot of uses this is probably an adequate entropy source - but I'm
not a cryptographer. If you'd like better entropy consider subclassing
and provding a C<random_data> that reads from /dev/urandom.

=cut

sub random_data {
    my $self = shift;
    return join( ':',
        time(), $self->{seed}++,
        refaddr( $self ),
        refaddr( \my $dummy ), $$ );
}

=head2 C<< passcodes >>

Get an array of passcodes.

    my @passcodes = $ppp->passcodes(1, 70, $seq_key);

The first two arguments are the starting position (1 .. n) and the
number of passcodes to generate.

Returns an array of strings containing the generated passcodes.

=cut

sub passcodes {
    croak "passcodes requires 3 args" unless @_ == 4;
    my ( $self, $first, $count, $sequence ) = @_;

    croak "Sequence must be 64 characters long"
      unless length( $sequence ) == 64;

    my @passcodes = ();

    croak "Starting index is 1" if $first <= 0;
    $first--;

    $first *= $count;

    my $codelen = $self->codelen;

    my $rij = Crypt::Rijndael->new( pack( 'H*', $sequence ),
        Crypt::Rijndael::MODE_ECB );

    while ( @passcodes < $count ) {
        my $pos     = $first * 8 * $codelen;
        my $n       = $pos / 128;
        my $offset  = $pos % 128;
        my $desired = int( $offset / 8 ) + $codelen;
        my $raw     = '';

        for my $j ( 0 .. 1 ) {
            my $n_bits = pack( "V*", "$n" );    # $n_bits .= ;
            $raw .= $rij->encrypt(
                $n_bits . "\0" x ( 16 - length( $n_bits ) % 16 ) );
            last if length( $raw ) >= $desired;
            $n++;
        }

        push @passcodes,
          $self->_alpha_encode( substr( $raw, $offset / 8, $codelen ),
            $codelen );

        $first++;
    }

    return @passcodes;
}

{
    my $GOT_BIGINT;

    sub _got_bigint {
        defined $GOT_BIGINT and return $GOT_BIGINT;
        return $GOT_BIGINT = eval 'use Math::BigInt; 1' ? 1 : 0;
    }
}

sub _alpha_encode {
    my ( $self, $data, $bytes ) = @_;
    my $code;

    if ( _got_bigint() && $bytes > 4 ) {
        # Make a big hex constant
        $code = Math::BigInt->new(
            '0x'
              . join( '',
                map { sprintf( "%02x", ord( $_ ) ) } reverse split //,
                $data )
        );
    }
    else {
        $code = unpack( 'V', $data . "\0" x ( 4 - length $data ) );
    }

    my $limit = 2**( $bytes * 8 );

    my @alphabet   = split //, $self->alphabet;
    my $code_space = @alphabet;
    my @out        = ();
    my $max        = 1;

    while ( $max < $limit ) {
        push @out, $alphabet[ $code % $code_space ];
        $code = int( $code / $code_space );
        $max *= $code_space;
    }

    return join '', @out;
}

1;
__END__

=head1 CONFIGURATION AND ENVIRONMENT
  
Crypt::PerfectPaperPasswords requires no configuration files or environment variables.

=head1 DEPENDENCIES

L<Crypt::Rijndael>

L<Digest::SHA256>

L<Scalar::Util>

L<Time::HiRes>

L<Math::BigInt> (optional)


=head1 INCOMPATIBILITIES

None reported.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-crypt-perfectpaperpasswords@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.

=head1 AUTHOR

Andy Armstrong  C<< <andy@hexten.net> >>

Original Perfect Paper Passwords implementation by Steve Gibson. More details
here:

    http://www.grc.com/ppp.htm

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2007, Andy Armstrong C<< <andy@hexten.net> >>.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.
